py/packages/genkit/src/genkit/ai/document.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""Document model for the Genkit framework.
+
+This module provides types for creating and managing multi-modal documents in
+Genkit.
+"""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import Any, ClassVar
+
+from genkit.core.typing import (
+    DocumentData,
+    DocumentPart,
+    Embedding,
+    MediaPart,
+    TextPart,
+)
+from genkit.core.typing import Media1 as MediaPartModel
+
+TEXT_DATA_TYPE: str = 'text'
+
+
+class Document(DocumentData):
+    """Represents document content along with its metadata.
+
+    This object can be embedded, indexed or retrieved. Each document can contain
+    multiple parts (for example text and an image).
+    """
+
+    def __init__(
+        self,
+        content: list[DocumentPart],
+        metadata: dict[str, Any] | None = None,
+    ) -> None:
+        """Initialize Document object."""
+        doc_content = deepcopy(content)
+        doc_metadata = deepcopy(metadata)
+        super().__init__(content=doc_content, metadata=doc_metadata)
+
+    @staticmethod
+    def from_text(
+        text: str, metadata: dict[str, Any] | None = None
+    ) -> Document:
+        """Construct a Document from a single text part."""
+        return Document(content=[TextPart(text=text)], metadata=metadata)
+
+    @staticmethod
+    def from_media(
+        url: str,
+        content_type: str | None = None,
+        metadata: dict[str, Any] | None = None,
+    ) -> Document:
+        """Construct a Document from a single media part."""
+        return Document(
+            content=[
+                MediaPart(
+                    media=MediaPartModel(url=url, content_type=content_type)
+                )
+            ],
+            metadata=metadata,
+        )
+
+    @staticmethod
+    def from_data(
+        data: str,
+        data_type: str | None = None,
+        metadata: dict[str, Any] | None = None,
+    ) -> Document:
+        """Construct a Document from a single media part."""
+        if data_type == TEXT_DATA_TYPE:
+            return Document.from_text(data, metadata)
+        return Document.from_media(data, data_type, metadata)
+
+    def text(self) -> str:
+        """Concatenates all `text` parts with no delimiter."""
+        text_parts = map(lambda part: part.root.text or '', self.content)
+        return ''.join(text_parts)
+
+    def media(self) -> list[MediaPartModel]:
+        """Media array getter."""
+        media_parts = map(lambda part: part.root.media, self.content)
```

Comment:
this might be a little shorter/cleaner:
```suggestion
        media_parts = part.root.media for part in self.content
```